File: ga/glm4_result_2024_6/result/1968D_glm/49_1968D.c
use std::io;
use std::cmp::min;
use std::cmp::max;

const N: usize = 200000;

fn main() {
    let mut t = 0;
    let mut n = 0;
    let mut k = 0;
    let mut a = 0;
    let mut b = 0;
    let mut arr = [0; N + 1];
    let mut aa = [0; N + 1];
    let mut p1 = 0;
    let mut p2 = 0;
    let mut p1_ = 0;
    let mut p2_ = 0;
    let mut i = -1;
    let mut i_ = -1;
    let mut j = -1;
    let mut j_ = -1;

    let stdin = io::stdin();
    let line = stdin.lock().lines().next().unwrap().unwrap();
    t = line.parse::<i32>().unwrap();

    while t > 0 {
        p1 = 0;
        p2 = 0;
        p1_ = 0;
        p2_ = 0;
        i = -1;
        i_ = -1;
        j = -1;
        j_ = -1;

        let line = stdin.lock().lines().next().unwrap().unwrap();
        let mut iter = line.split_whitespace();
        n = iter.next().unwrap().parse::<i32>().unwrap() as usize;
        k = iter.next().unwrap().parse::<i32>().unwrap() as usize;
        a = iter.next().unwrap().parse::<i32>().unwrap() as usize;
        b = iter.next().unwrap().parse::<i32>().unwrap() as usize;

        for x in 1..=n {
            let line = stdin.lock().lines().next().unwrap().unwrap();
            arr[x] = line.parse::<i32>().unwrap();
        }

        for x in 1..=n {
            let line = stdin.lock().lines().next().unwrap().unwrap();
            aa[x] = line.parse::<i32>().unwrap();
        }

        for x in 1..=min(n, k) {
            p1 = max(p1, aa[a] as i64 * (k - x + 1) as i64 + p1_);
            p1_ += aa[a];
            a = arr[a];

            p2 = max(p2, aa[b] as i64 * (k - x + 1) as i64 + p2_);
            p2_ += aa[b];
            b = arr[b];
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }

        t -= 1;
    }
}

File: ga/gpt3_5_result_2024_6/result/1968D_gpt/49_1968D.rs
use std::io::{self, BufRead};

fn min(a: i32, b: i32) -> i32 {
    if a < b { a } else { b }
}

fn maxl(a: i64, b: i64) -> i64 {
    if a > b { a } else { b }
}

fn main() {
    let stdin = io::stdin();
    let mut input = stdin.lock().lines().map(|line| line.unwrap());

    let t: i32 = input.next().unwrap().trim().parse().unwrap();
    
    for _ in 0..t {
        let mut p1: i64 = 0;
        let mut p2: i64 = 0;
        let mut p1_: i64 = 0;
        let mut p2_: i64 = 0;
        let mut i: i32 = -1;
        let mut i_: i32 = -1;
        let mut j: i32 = -1;
        let mut j_: i32 = -1;

        let line: String = input.next().unwrap().trim().to_string();
        let mut parts = line.split_whitespace();
        let n: i32 = parts.next().unwrap().parse().unwrap();
        let k: i32 = parts.next().unwrap().parse().unwrap();
        let a: i32 = parts.next().unwrap().parse().unwrap();
        let b: i32 = parts.next().unwrap().parse().unwrap();
        let m = min(n, k);
        let mut a_: i32 = a;
        let mut b_: i32 = b;

        let mut arr: Vec<i32> = vec![0; (n + 1) as usize];
        let mut aa: Vec<i32> = vec![0; (n + 1) as usize];

        let line1: String = input.next().unwrap().trim().to_string();
        let parts1: Vec<&str> = line1.split_whitespace().collect();
        for (index, part) in parts1.iter().enumerate() {
            arr[index + 1] = part.parse().unwrap();
        }

        let line2: String = input.next().unwrap().trim().to_string();
        let parts2: Vec<&str> = line2.split_whitespace().collect();
        for (index, part) in parts2.iter().enumerate() {
            aa[index + 1] = part.parse().unwrap();
        }

        for x in 1..=m {
            p1 = maxl(p1, (aa[a as usize] * (k - x + 1) as i64) + p1_);
            p1_ += aa[a as usize] as i64;
            a = arr[a as usize];
            p2 = maxl(p2, (aa[b as usize] * (k - x + 1) as i64) + p2_);
            p2_ += aa[b as usize] as i64;
            b = arr[b as usize];
        }

        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}